lib/spack/spack/cmd/container.py

```diff
@@ -0,0 +1,151 @@
+        '--multi-stage', action='store_true', default=False,
+        help='generate a multi-stage Dockerfile (Docker specific)')
+    subparser.add_argument(
+        '--no-checksum', action='store_true', default=False,
+        help='do not use checksums to verify downloaded files (unsafe)')
+    subparser.add_argument(
+        '--show-log-on-error', action='store_true', default=False,
+        help='print full build log if build fails')
+    subparser.add_argument(
+        '--spack-branch', '--branch', action='store', type=str,
+        default='master',
+        help='branch of Spack to deploy (default: master)')
+
+
+def container(parser, args):
+    if not args.specs:
+        tty.die("spack spec requires at least one spec")
+
+    recipe = hpccm.Stage()
+
+    recipe += baseimage(image=args.baseimage, _distro=args.distro)
+
+    # OS dependencies
+    recipe += hpccm.building_blocks.packages(
+        apt=['autoconf', 'build-essential', 'bzip2', 'ca-certificates',
+             'coreutils', 'cpio', 'curl', 'environment-modules', 'git',
+             'gzip', 'libssl-dev', 'make', 'openssh-client', 'patch',
+             'pkg-config', 'svn', 'tar', 'tcl', 'unzip', 'zlib1g'],
+        yum=['autoconf', 'bzip2', 'ca-certificates', 'coreutils', 'cpio',
+             'curl', 'environment-modules', 'git', 'gzip', 'make',
+             'openssh-clients', 'openssl-devel', 'patch', 'pkg-config',
+             'svn', 'tar', 'tcl', 'unzip', 'zlib-devel'])
+
+    # GNU compilers
+    recipe += hpccm.building_blocks.gnu()
+
+    # Spack itself
+    spack_bb = bb_base()
+    spack_bb += comment('Spack')
+    spack_bb += shell(commands=[
+        'mkdir -p /opt && cd /opt',
+        'git clone --depth=1 --branch {} https://github.com/spack/spack'.format(args.spack_branch),
+        '/opt/spack/bin/spack bootstrap',
+        'ln -s /opt/spack/share/spack/setup-env.sh /etc/profile.d/spack.sh',
+        'ln -s /opt/spack/share/spack/spack-completion.bash /etc/profile.d'])
+    spack_bb += environment(variables={'PATH': '/opt/spack/bin:$PATH',
+                                       'FORCE_UNSAFE_CONFIGURE': '1'})
+    recipe += spack_bb
+
+    # spack install command line arguments
+    install_args = ['-y']
+    if args.no_checksum:
+        install_args.append('--no-checksum')
+    if args.show_log_on_error:
+        install_args.append('--show-log-on-error')
+    install_args_str = ' '.join(install_args)
+
+    # Process each spec
+    for spec in spack.cmd.parse_specs(args.specs):
+        spec.concretize()
+
+        bb = bb_base()
+        bb += comment('{0} {1}'.format(spec.name, spec.version))
+
+        # Process each dependency.  Use a selectable depth to balance
+        # the build cache vs. container spec file complexity
+        for depth, dspec in spec.traverse_edges(order='post', depth=True):
+            if depth <= args.depth:
+                node = dspec.spec
+                cmd = 'spack install {0} {1}@{2}'.format(install_args_str,
```

Comment:
Sorry, I'm not sure I fully understand this comment.  The code calls `spec.concretize()`?

I changed the output to include the Spack package options (which was probably an earlier oversight - sorry!).  The output is now:

```
$ spack container gromacs@2019+cuda
[snip]
RUN spack install -y cuda@10.0.130 && \
    spack clean --all
RUN spack install -y openmpi@3.1.3~cuda+cxx_exceptions fabrics=auto ~java~legacylaunchers~memchecker~pmi schedulers=auto ~sqlite3~thread_multiple+vt && \
    spack clean --all
RUN spack install -y fftw@3.3.8+double+float~fma+long_double+mpi~openmp~pfft_patches~quad simd=avx,avx2,sse2 && \
    spack clean --all
RUN spack install -y gromacs@2019 build_type=RelWithDebInfo +cuda~double+mpi~plumed+rdtscp+shared simd=auto && \
    spack clean --all
```

OpenMPI is installed without CUDA, `openmpi@3.1.3~cuda+cxx_exceptions`.  Is the problem that `spack install gromacs ... +cuda...` will cause a second version of OpenMPI to be installed?  The output of `spack spec gromacs@2019+cuda` matches what is shown above, so I'm not sure?  

I did not use the `--only` option of `spack install` because not every node from the DAG is explicitly made into a container layer (see `--depth`).  My understanding is that this would fail otherwise?  (The idea behind `--depth` is to strike a balance between the Docker build cache, the number of layers, and Spack).

If I understand the yaml file suggestion correctly, it would invalidate the use of the Docker build cache (the install would be a single command), which would be undesirable.

Reply:
> If I understand the yaml file suggestion correctly, it would invalidate the use of the Docker build cache (the install would be a single command), which would be undesirable.

You can write the full spec DAG into a set of YAML files: the loop would be as you have it here, but instead of writing out a spack invocation command, you can generate the spack yaml for the dependency.

> I changed the output to include the Spack package options

This will generally work but still can run into problems: imagine a hierarchy like `X->Y->Z` where

* X depends on Z+foo
* Y also depends on Z but doesn't require +foo
* Z builds ~foo by default

If you separately concretize Y, you will get `Y->Z~foo` but when concretizing X it will be `X->Y->Z+foo`, so two instances of Z will be built.

If you save the spec YAMLs you can use `--only=package` on everything above the truncation point and don't specify `--only` for the depth where you want to install everything below.

Reply:
Can you please point me to an actual example of a package that demonstrates the above dependency scenario?  I'm looking for a test case to ensure that this is handled correctly (and to make sure I fully understand the issue).  Thanks.

Reply:
See `dealii+int64`, `suite-sparse`, and `metis`: the `metis` dependency from concretizing `suite-sparse` (specifically, if you take the `suite-sparse` line from `spack spec dealii+int64`, it will lack the `+int64` for `metis`, but `metis` in the `dealii+int64` DAG will have `+int64`.

In general, dependencies can acquire constraints from multiple different dependents, so concretizing any set of specs in isolation from one another (which happens as part of `spack install` when you supply a command-line spec) may result in some form of this issue.

Reply:
Thanks for the example.  It's very useful.

It appears that `--only` is ignored when installing from a yaml spec?

A significant issue when generating a yaml spec is that the spec includes compiler and arch sections for the system on which the spec is concretized.  Since the container environment is likely to be different than the host where `spack container` is run, the yaml spec generated on the host cannot be used in the container (unlike the current implementation).  For example, the host might be CentOS 7 with the distro GNU 4.8 compilers, while the container could be Ubuntu 16 with the distro GNU 5.4 compilers.

The answer might seem to be generate the spec in the container (`RUN spack spec --yaml ... > file.yml`).  But by doing that, the container build cache can not be reused (`gromacs@2019` can not reuse the build cache from `gromacs@2018.2`), which is one of the main reasons for going through the trouble of iterating over the DAG in the first place.  

(Even solving the compiler/arch issue, generating the yaml spec on the host and copying it into the container at build time would similarly invalidate the container build cache.  I don't see a way to effectively use both yaml specs and the container build cache.)

If that's the case, the remaining benefit of using the container build cache to recover from transient build failures may not be enough to warrant the complexity of working with the DAG?

Reply:
Unrolling the concretized spec has been removed.  It can be re-enabled in a future PR if Spack itself is enhanced to support this use case.